oss_src/unity/extensions/additional_sframe_utilities.cpp

```diff
@@ -18,28 +19,59 @@ typedef int(*sframe_callback_type)(const graphlab::flexible_type*, size_t, void*
 
 void sarray_callback(graphlab::gl_sarray input, size_t callback_addr, size_t callback_data_addr,
                      size_t begin, size_t end) {
+  gl_sarray sliced_input = input[{(int64_t)begin, (int64_t)end}];
+
+  // materialize in parallel to rows_by_segment_id
+  size_t nthreads = thread::cpu_count();
+  std::vector<std::vector<std::vector<flexible_type>>> rows_by_segment_id;
+  rows_by_segment_id.resize(nthreads);
+  auto copy_out_callback = [&](size_t segment_id,
+                               const std::shared_ptr<sframe_rows>& data) {
+    for (auto& row: (*data)) {
+      rows_by_segment_id[segment_id].push_back(row);
+    }
+    return false;
+  };
+  sliced_input.materialize_to_callback(copy_out_callback, nthreads);
```

Comment:
Agreed. However, the benefit is that the materialization is now done in parallel with out requiring the callback function to be reentrent. Since the callback is ranged, the risk of out-of-memory is limited and controllable.